src/interpreter/utils.ts

```diff
@@ -124,3 +152,100 @@ export const setParamArgs = (
   }
   return nameValueMappings;
 };
+
+const typeOf = (v: Value): string => {
+  if (v === null) {
+    return 'null';
+  } else if (Array.isArray(v)) {
+    return 'array';
+  } else {
+    return typeof v;
+  }
+};
+
+const isNumber = (v: Value): v is number => typeOf(v) === 'number';
+const isString = (v: Value): v is string => typeOf(v) === 'string';
+
+export const checkBinaryOperation = (
+  operator: BinaryOperator,
+  left: Value,
+  right: Value
+): void => {
+  switch (operator) {
+    case '-':
+    case '*':
+    case '/':
+    case '%':
+      if (!isNumber(left)) {
+        throw new TypeError('number', typeOf(left), TypeErrorSide.LHS);
+      }
+      if (!isNumber(right)) {
+        throw new TypeError('number', typeOf(right), TypeErrorSide.RHS);
+      }
+      break;
+    case '+':
+    case '<':
+    case '<=':
+    case '>':
+    case '>=':
+      if (!isNumber(left) && !isString(left)) {
+        throw new TypeError(
+          'string or number',
+          typeOf(left),
+          TypeErrorSide.LHS
+        );
+      }
+      if (isNumber(left) && !isNumber(right)) {
+        throw new TypeError('number', typeOf(right), TypeErrorSide.RHS);
+      }
+      if (isString(left) && !isString(right)) {
+        throw new TypeError('string', typeOf(right), TypeErrorSide.RHS);
+      }
+      break;
+    default:
+  }
+};
+
+export function evaluateBinaryExpression(
+  operator: BinaryOperator,
+  left: any,
+  right: any
+): Value {
+  switch (operator) {
+    case '+':
+      // Separate typeguards are required due to typescript rules: https://typescript-eslint.io/rules/restrict-plus-operands/
```

Comment:
Yes, it just looks funny, because the `left + right` is duplicated. But I cannot write only a single predicate `(isNumber(left) && isNumber(right)) || (isString(left) && isString(right))`